File: frontend/src/api/size.js
/**
 * Size API
 * Functions for size prediction and size chart endpoints
 */

import { ENDPOINTS } from './config'
import { apiPost, apiGet } from './client'

/**
 * Predict size based on body measurements
 *
 * @param {Object} measurements - Body measurements
 * @param {number} measurements.height - Height in cm
 * @param {number} measurements.chest - Chest circumference in cm
 * @param {number} measurements.waist - Waist circumference in cm
 * @param {number} measurements.hips - Hip circumference in cm
 * @param {number} measurements.shoulder_width - Shoulder width in cm
 * @param {Object} options - Additional options
 * @param {Object} options.sizeChart - Optional brand-specific size chart
 * @param {boolean} options.useStandardChart - Use standard chart (default: true)
 * @returns {Promise<Object>} - Size prediction response
 */
export async function predictSize(measurements, options = {}) {
  const {
    sizeChart = null,
    useStandardChart = true,
    category = 'shirts',
    gender = 'men'
  } = options

  // Validate measurements - ensure no null/undefined values
  const validMeasurements = {
    height: Number(measurements.height) || 0,
    chest: Number(measurements.chest) || 0,
    waist: Number(measurements.waist) || 0,
    hips: Number(measurements.hips) || 0,
    shoulder_width: Number(measurements.shoulder_width) || 0
  }

  console.log("Predict payload - measurements:", validMeasurements)
  console.log("Predict payload - options:", { category, gender, sizeChart, useStandardChart })

  // Only include category/gender if not using a custom size chart
  const requestBody = {
    measurements: validMeasurements,
    size_chart: sizeChart,
    use_standard_chart: useStandardChart,
    ...(sizeChart ? {} : { category, gender })
  }

  console.log("Predict request body:", JSON.stringify(requestBody))

  return apiPost(ENDPOINTS.SIZE_PREDICT, requestBody)
}

/**
 * Get available standard size charts
 * @returns {Promise<Object>} - Available charts information
 */
export async function getStandardCharts() {
  return apiGet(ENDPOINTS.SIZE_STANDARD_CHARTS)
}

/**
 * Get a specific size chart
 *
 * @param {string} category - Category (shirts, pants, dresses, jackets)
 * @param {string} gender - Gender (men, women)
 * @returns {Promise<Object>} - Size chart data
 */
export async function getSizeChart(category, gender = 'men') {
  const params = new URLSearchParams({ gender })
  return apiGet(`${ENDPOINTS.SIZE_CHART}/${category}?${params.toString()}`)
}

/**
 * Validate if current size matches body measurements
 *
 * @param {Object} measurements - Body measurements
 * @param {string} currentSize - User's current size
 * @param {Object} sizeChart - Size chart to validate against
 * @returns {Promise<Object>} - Validation response
 */
export async function validateSize(measurements, currentSize, sizeChart) {
  return apiPost(ENDPOINTS.SIZE_VALIDATE, {
    measurements,
    current_size: currentSize,
    size_chart: sizeChart,
  })
}

/**
 * Submit fit feedback for model improvement
 *
 * @param {Object} measurements - Body measurements
 * @param {string} category - Garment category
 * @param {string} size - Size worn
 * @param {string} fitRating - Fit rating (too_tight, tight, perfect, loose, too_loose)
 * @returns {Promise<Object>} - Feedback response
 */
export async function submitFitFeedback(measurements, category, size, fitRating) {
  const validRatings = ['too_tight', 'tight', 'perfect', 'loose', 'too_loose']

  if (!validRatings.includes(fitRating)) {
    throw new Error(`Invalid fit rating. Must be one of: ${validRatings.join(', ')}`)
  }

  return apiPost('/size/feedback', {
    measurements,
    category,
    size,
    fit_rating: fitRating,
  })
}

File: backend/app/routers/size.py
"""
Size Prediction Router - API endpoints for size prediction and validation
"""

from fastapi import APIRouter, HTTPException
from app.models.schemas import (
    SizePredictionRequest,
    SizePredictionResponse,
    SizeValidationRequest,
    SizeValidationResponse
)
from app.services import chart_matcher, fit_model

router = APIRouter()


@router.post("/predict", response_model=SizePredictionResponse)
async def predict_size(request: SizePredictionRequest):
    """
    Predict the best size based on body measurements and size chart.

    - **measurements**: Body measurements (height, chest, waist, hips, shoulder_width)
    - **size_chart**: Optional brand-specific size chart
    - **use_standard_chart**: Use standard size chart if no brand chart provided
    """
    try:
        category = (
            request.size_chart.category
            if request.size_chart and request.size_chart.category
            else (request.category or "shirts")
        )
        gender = (
            request.size_chart.gender
            if request.size_chart and request.size_chart.gender
            else (request.gender or "men")
        )

        result = chart_matcher.predict_size(
            measurements=request.measurements,
            size_chart=request.size_chart,
            use_standard_chart=request.use_standard_chart,
            category=category,
            gender=gender,
            measurement_confidence=request.confidence
        )

        return result

    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Prediction error: {str(e)}")


@router.post("/validate", response_model=SizeValidationResponse)
async def validate_size(request: SizeValidationRequest):
    """
    Validate if user's current size matches their body measurements.

    - **measurements**: User's body measurements
    - **current_size**: Size user currently wears
    - **size_chart**: Size chart to validate against
    """
    try:
        result = chart_matcher.validate_size(request)
        return result

    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Validation error: {str(e)}")


@router.get("/standard-charts")
async def get_standard_charts():
    """Get list of available standard size charts."""
    return {
        "categories": list(chart_matcher.STANDARD_CHARTS.keys()),
        "genders": ["men", "women"],
        "charts": {
            "shirts": {
                "men": len(chart_matcher.STANDARD_CHARTS["shirts"]["men"]),
                "women": len(chart_matcher.STANDARD_CHARTS["shirts"]["women"])
            },
            "pants": {
                "men": len(chart_matcher.STANDARD_CHARTS["pants"]["men"]),
                "women": len(chart_matcher.STANDARD_CHARTS["pants"]["women"])
            },
            "dresses": {
                "women": len(chart_matcher.STANDARD_CHARTS["dresses"]["women"])
            },
            "jackets": {
                "men": len(chart_matcher.STANDARD_CHARTS["jackets"]["men"]),
                "women": len(chart_matcher.STANDARD_CHARTS["jackets"]["women"])
            }
        }
    }


@router.get("/chart/{category}")
async def get_size_chart(category: str, gender: str = "men"):
    """Get a specific standard size chart."""
    chart = chart_matcher.get_standard_chart(category, gender)

    if not chart:
        raise HTTPException(
            status_code=404,
            detail=f"No standard chart found for category '{category}' and gender '{gender}'"
        )

    return chart


@router.post("/feedback")
async def add_fit_feedback(
    measurements: dict,
    category: str,
    size: str,
    fit_rating: str
):
    """
    Submit fit feedback to improve recommendations.

    - **measurements**: Body measurements
    - **category**: Garment category
    - **size**: Size worn
    - **fit_rating**: Fit rating (too_tight, tight, perfect, loose, too_loose)
    """
    valid_ratings = ["too_tight", "tight", "perfect", "loose", "too_loose"]

    if fit_rating not in valid_ratings:
        raise HTTPException(
            status_code=400,
            detail=f"Invalid fit_rating. Must be one of: {valid_ratings}"
        )

    success = fit_model.add_feedback(measurements, category, size, fit_rating)

    if not success:
        raise HTTPException(
            status_code=500,
            detail="Failed to save feedback"
        )

    return {
        "success": True,
        "message": "Feedback saved. Will be used to improve future recommendations."
    }


@router.post("/model/train")
async def train_fit_model(epochs: int = 50, batch_size: int = 16):
    """Train the fit prediction model on accumulated feedback."""
    result = fit_model.train_model(epochs=epochs, batch_size=batch_size)

    if not result["success"]:
        raise HTTPException(status_code=400, detail=result["message"])

    return result


@router.get("/model/stats")
async def get_model_stats():
    """Get training data statistics and model info."""
    return fit_model.get_training_stats()


@router.post("/predict-fit")
async def predict_fit_score(measurements: dict, category: str, size: str):
    """
    Predict fit score for a specific size using the ML model.

    - **measurements**: Body measurements
    - **category**: Garment category
    - **size**: Size to predict fit for
    """
    result = fit_model.predict_fit(measurements, category, size)

    if not result["success"]:
        raise HTTPException(status_code=500, detail=result["message"])

    return result
